The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7095000634/job/19311231994) [(plain)](https://github.com/rust-lang/rust/commit/0cd4e1880741249dfb06f8f1787721c258610cd8/checks/19311231994/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
     |
1672 |         if let Some(stack_protector) = self.config.rust_stack_protector {
     |                ^^^^^^^^^^^^^^^^^^^^^   -------------------------------- this expression has type `std::string::String`
     |                |
     |                expected `String`, found `Option<_>`
     = note: expected struct `std::string::String`
                  found enum `std::option::Option<_>`

error[E0308]: mismatched types
error[E0308]: mismatched types
##[error]    --> src/core/builder.rs:1673:27
     |
1673 |             rustflags.arg(format!("-Zstack-protector={stack_protector}"));
     |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `&str`, found `String`
     = note: this error originates in the macro `format` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0308`.
error: could not compile `bootstrap` (lib) due to 2 previous errors
---
     |
1672 |         if let Some(stack_protector) = self.config.rust_stack_protector {
     |                ^^^^^^^^^^^^^^^^^^^^^   -------------------------------- this expression has type `std::string::String`
     |                |
     |                expected `String`, found `Option<_>`
     = note: expected struct `std::string::String`
                  found enum `std::option::Option<_>`

error[E0308]: mismatched types
error[E0308]: mismatched types
##[error]    --> src/core/builder.rs:1673:27
     |
1673 |             rustflags.arg(format!("-Zstack-protector={stack_protector}"));
     |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `&str`, found `String`
     = note: this error originates in the macro `format` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0308`.
error: could not compile `bootstrap` (lib) due to 2 previous errors
---
     |
1672 |         if let Some(stack_protector) = self.config.rust_stack_protector {
     |                ^^^^^^^^^^^^^^^^^^^^^   -------------------------------- this expression has type `std::string::String`
     |                |
     |                expected `String`, found `Option<_>`
     = note: expected struct `std::string::String`
                  found enum `std::option::Option<_>`

error[E0308]: mismatched types
error[E0308]: mismatched types
##[error]    --> src/core/builder.rs:1673:27
     |
1673 |             rustflags.arg(format!("-Zstack-protector={stack_protector}"));
     |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `&str`, found `String`
     = note: this error originates in the macro `format` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0308`.
error: could not compile `bootstrap` (lib) due to 2 previous errors
---
     |
1672 |         if let Some(stack_protector) = self.config.rust_stack_protector {
     |                ^^^^^^^^^^^^^^^^^^^^^   -------------------------------- this expression has type `std::string::String`
     |                |
     |                expected `String`, found `Option<_>`
     = note: expected struct `std::string::String`
                  found enum `std::option::Option<_>`

error[E0308]: mismatched types
error[E0308]: mismatched types
##[error]    --> src/core/builder.rs:1673:27
     |
1673 |             rustflags.arg(format!("-Zstack-protector={stack_protector}"));
     |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `&str`, found `String`
     = note: this error originates in the macro `format` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0308`.
error: could not compile `bootstrap` (lib) due to 2 previous errors
---
     |
1672 |         if let Some(stack_protector) = self.config.rust_stack_protector {
     |                ^^^^^^^^^^^^^^^^^^^^^   -------------------------------- this expression has type `std::string::String`
     |                |
     |                expected `String`, found `Option<_>`
     = note: expected struct `std::string::String`
                  found enum `std::option::Option<_>`

error[E0308]: mismatched types
error[E0308]: mismatched types
##[error]    --> src/core/builder.rs:1673:27
     |
1673 |             rustflags.arg(format!("-Zstack-protector={stack_protector}"));
     |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `&str`, found `String`
     = note: this error originates in the macro `format` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0308`.
error: could not compile `bootstrap` (lib) due to 2 previous errors
```

</details>

--- a/config.example.toml
+++ b/config.example.toml
@@ -600,6 +600,16 @@ change-id = 117813
 # desired in distributions, for example.
 #rpath = true
 
+# Indicates whether symbols should be stripped using `-Cstrip=symbols`.
+#strip = false
+
+# Indicates whether stack protectors should be used
+# via the unstable option `-Zstack-protector`.
+#
+# Valid options are : `none`(default),`basic`,`strong`, or `all`.
+# `strong` and `basic` options may be buggy and are not recommended, see rust-lang/rust#114903.
+#stack-protector = "none"
+
 # Prints each test name as it is executed, to help debug issues in the test harness itself.
 #verbose-tests = false
 

--- a/src/bootstrap/src/core/builder.rs
+++ b/src/bootstrap/src/core/builder.rs
@@ -1667,6 +1667,12 @@ impl<'a> Builder<'a> {
             }
         }
 
+        cargo.env(profile_var("STRIP"), self.config.rust_strip.to_string());
+
+        if let Some(stack_protector) = &self.config.rust_stack_protector {
+            rustflags.arg(&format!("-Zstack-protector={stack_protector}"));
+        }
+
         if let Some(host_linker) = self.linker(compiler.host) {
             hostflags.arg(format!("-Clinker={}", host_linker.display()));
         }

--- a/src/bootstrap/src/core/config/config.rs
+++ b/src/bootstrap/src/core/config/config.rs
@@ -222,6 +222,8 @@ pub struct Config {
     pub rust_debuginfo_level_tests: DebuginfoLevel,
     pub rust_split_debuginfo: SplitDebuginfo,
     pub rust_rpath: bool,
+    pub rust_strip: bool,
+    pub rust_stack_protector: Option<String>,
     pub rustc_parallel: bool,
     pub rustc_default_linker: Option<String>,
     pub rust_optimize_tests: bool,
@@ -1001,6 +1003,8 @@ define_config! {
         description: Option<String> = "description",
         musl_root: Option<String> = "musl-root",
         rpath: Option<bool> = "rpath",
+        strip: Option<bool> = "strip",
+        stack_protector: Option<String> = "stack-protector",
         verbose_tests: Option<bool> = "verbose-tests",
         optimize_tests: Option<bool> = "optimize-tests",
         codegen_tests: Option<bool> = "codegen-tests",
@@ -1069,6 +1073,7 @@ impl Config {
         config.docs = true;
         config.docs_minification = true;
         config.rust_rpath = true;
+        config.rust_strip = false;
         config.channel = "dev".to_string();
         config.codegen_tests = true;
         config.rust_dist_src = true;
@@ -1422,6 +1427,8 @@ impl Config {
             set(&mut config.rust_optimize_tests, rust.optimize_tests);
             set(&mut config.codegen_tests, rust.codegen_tests);
             set(&mut config.rust_rpath, rust.rpath);
+            set(&mut config.rust_strip, rust.strip);
+            config.rust_stack_protector = rust.stack_protector;
             set(&mut config.jemalloc, rust.jemalloc);
             set(&mut config.test_compare_mode, rust.test_compare_mode);
             set(&mut config.backtrace, rust.backtrace);

